thinc/loss.py

```diff
@@ -58,142 +171,190 @@ def __init__(
+        missing = []
+        negatives_mask = xp.ones((len(truths), len(self.names)), dtype="f")
+        truths_int = []
+        for i, value in enumerate(truths):
+            if not isinstance(value, str):
+                raise ValueError(
+                    "The first value of the truths was of type "
+                    f"string, but found {type(value)} during iteration."
+                )
+            # missing value
+            if value == missing_value:
+                label_i = self._name_to_i[self.names[0]]
+                missing.append(i)
+            # negative labels
+            elif self.neg_prefix and value.startswith(self.neg_prefix):
+                label_i = self._name_to_i[value[len(self.neg_prefix) :]]
+                negatives_mask[i] = 0  # type: ignore
+                negatives_mask[i][label_i] = -1  # type: ignore
+            # nothing special
+            else:
+                label_i = self._name_to_i[value]
+            truths_int.append(label_i)
+        xp_truths = cast(Ints1d, xp.asarray(truths_int, dtype="i"))
+        truths_2d = to_categorical(
+            xp_truths, n_classes=guesses.shape[-1], label_smoothing=self.label_smoothing
+        )
+        mask = _make_mask(guesses, missing)
+        truths_2d *= negatives_mask
+        truths_2d[truths_2d == -1] = 0
+        negatives_mask[negatives_mask == -1] = 1
+        mask *= negatives_mask
+        return cast(Floats2d, truths_2d), mask
+
+    def convert_truths(
+        self, truths: Categories1d, guesses: Floats2d
+    ) -> Tuple[Floats2d, Floats2d]:
+        guesses_f2d = _to_array(guesses)
+
+        if is_xp_array(truths):
+            self._check_ints1d(truths)
+            xp_truths = cast(Ints1d, truths)
+            truths_2d = to_categorical(
+                xp_truths,
                 label_smoothing=self.label_smoothing,
+                n_classes=guesses_f2d.shape[1],
             )
-        else:
-            if self.label_smoothing:
+            mask = _make_mask_by_value(truths_2d, guesses_f2d, self.missing_value)
+        elif isinstance(truths, Sequence):
+            if isinstance(truths[0], int):
+                truths_2d, mask = self._convert_ints(
+                    guesses_f2d, cast(Sequence[int], truths)
+                )
+            elif isinstance(truths[0], str):
+                truths_2d, mask = self._convert_strs(
+                    guesses_f2d, cast(Sequence[str], truths)
+                )
+            else:
                 raise ValueError(
-                    "Label smoothing is only applied, when truths have type "
-                    "List[str], List[int] or Ints1d, but it seems like Floats2d "
-                    "was provided."
+                    "When truths to SparseCE is provided "
+                    "in Sequence format, elements need to be "
+                    "of type str or int, but first element "
+                    f"was found to be {type(truths[0])}."
                 )
```

Comment:
We should check all elements of `truths` like we do in `_convert_strs`.

Reply:
In this block we only check the first element and then `_convert_ints` and `_convert_strs` check all when they run and complain that the first one was of some type but now i ran into something else. So I think all elements get eventually checked here. No?

Reply:
Ah, good point. Didn't notice that.